rust/frontend/src/ndarray.rs

```diff
@@ -0,0 +1,359 @@
+//! This module implements [`NDArray`] type for working with TVM tensors or
+//! covertsion from Rust's ndarray to TVM `NDArray`.
+//!
+//! One can create an empty NDArray given the shape, context and dtype using [`empty`].
+//! To create an NDArray from a mutable buffer in cpu use [`copy_from_buffer`].
+//! To copy an NDArray to different context use [`copy_to_ctx`].
+//!
+//! Given a [`Rust's dynamic ndarray`], one can convert it to TVM NDArray as follows:
+//!
+//! # Example
+//!
+//! ```
+//! let a = Array::from_shape_vec((2, 2), vec![1f32, 2., 3., 4.])
+//!     .unwrap()
+//!     .into_dyn(); // Rust's ndarray
+//! let nd = NDArray::from_rust_ndarray(&a, TVMContext::cpu(0), TVMType::from("float")).unwrap();
+//! assert_eq!(nd.shape(), Some(vec![2, 2]));
+//! let rnd: ArrayD<f32> = ArrayD::try_from(&nd).unwrap();
+//! assert!(rnd.all_close(&a, 1e-8f32));
+//! ```
+//!
+//! [`Rust's dynamic ndarray`]:https://docs.rs/ndarray/0.12.1/ndarray/
+//! [`copy_from_buffer`]:struct.NDArray.html#method.copy_from_buffer
+//! [`copy_to_ctx`]:struct.NDArray.html#method.copy_to_ctx
+
+use std::convert::TryFrom;
+use std::mem;
+use std::os::raw::c_int;
+use std::ptr;
+use std::slice;
+
+use num_traits::Num;
+use rust_ndarray::{Array, ArrayD};
+
+use ts;
+
+use Error;
+use Result;
+use TVMByteArray;
+use TVMContext;
+use TVMType;
+
+/// See the [`module-level documentation`](../ndarray/index.html) for more details.
+///
+/// Wrapper around TVM array handle.
+#[derive(Debug)]
+pub struct NDArray {
+    pub(crate) handle: ts::TVMArrayHandle,
+    is_view: bool,
+}
+
+impl NDArray {
+    pub(crate) fn new(handle: ts::TVMArrayHandle, is_view: bool) -> Self {
+        NDArray {
+            handle: handle,
+            is_view: is_view,
+        }
+    }
+
+    /// Returns the underlying array handle.
+    #[inline]
+    pub fn as_handle(&self) -> ts::TVMArrayHandle {
+        self.handle
+    }
+
+    #[inline]
+    pub fn is_view(&self) -> bool {
+        self.is_view
+    }
+
+    /// Returns the shape of the NDArray.
+    pub fn shape(&self) -> Option<Vec<usize>> {
+        let arr = unsafe { *(self.handle) };
+        if arr.shape.is_null() || arr.data.is_null() {
+            return None;
+        };
+        let slc = unsafe { slice::from_raw_parts_mut(arr.shape as *mut usize, arr.ndim as usize) };
+        Some(slc.to_vec())
+    }
+
+    /// Returns the total number of entries of the NDArray.
+    #[inline]
+    pub fn size(&self) -> Option<usize> {
+        self.shape().map(|v| v.into_iter().product())
```

Comment:
Because in `empty`, other methods depending on `empty` such as `to_vec`, they need a mutable reference to vec. So I think a reference wouldn't be enough!

Reply:
I'm more worried that any constructor wants a `mut &Vec` from your calling `size()`. I would have `size()` return a slice and then use `to_vec` before passing ownership of the vec to whatever constructor wants it. `mut _` is a good thing to avoid.

Reply:
Sorry, my earlier comment was meant for the previous of line 78. Please [see](https://github.com/dmlc/tvm/pull/2292#discussion_r241923970).